src/Features/Lsif/Generator/CompilerInvocation.cs

```diff
@@ -0,0 +1,184 @@
+                return DocumentInfo.Create(
+                    DocumentId.CreateNewId(projectId, mappedPath),
+                    name: mappedPath,
+                    filePath: mappedPath,
+                    loader: new FileTextLoader(mappedPath, parsedCommandLine.Encoding));
+            }
+
+            var projectInfo = ProjectInfo.Create(
+                projectId,
+                VersionStamp.Default,
+                name: Path.GetFileNameWithoutExtension(invocationInfo.ProjectFilePath),
+                assemblyName: parsedCommandLine.CompilationName!,
+                language: languageName,
+                filePath: invocationInfo.ProjectFilePath,
+                outputFilePath: parsedCommandLine.OutputFileName,
+                parsedCommandLine.CompilationOptions,
+                parsedCommandLine.ParseOptions,
+                parsedCommandLine.SourceFiles.Select(s => CreateDocumentInfo(unmappedPath: s.Path)),
+                metadataReferences: parsedCommandLine.MetadataReferences.Select(r => MetadataReference.CreateFromFile(mapPath(r.Reference), r.Properties)),
+                additionalDocuments: parsedCommandLine.AdditionalFiles.Select(f => CreateDocumentInfo(unmappedPath: f.Path)))
+                .WithAnalyzerConfigDocuments(parsedCommandLine.AnalyzerConfigPaths.Select(CreateDocumentInfo));
+
+            workspace.AddProject(projectInfo);
+
+            var compilation = await workspace.CurrentSolution.GetProject(projectId)!.GetRequiredCompilationAsync(CancellationToken.None);
+
+            return new CompilerInvocation(compilation, languageServices, invocationInfo.ProjectFilePath);
+        }
+
+        private static string GetLanguageName(CompilerInvocationInfo invocationInfo)
+        {
+            return invocationInfo.Tool switch
+            {
+                "csc" => LanguageNames.CSharp,
+                "vbc" => LanguageNames.VisualBasic,
+                _ => throw new NotSupportedException($"Tool {invocationInfo.Tool} is not supported."),
+            };
+        }
+
+        /// <summary>
+        /// Given the JSON description, returns a function that will map paths from the original paths to the current paths.
+        /// </summary>
+        private static Func<string, string> GetPathMapper(CompilerInvocationInfo invocationInfo)
+        {
+            return unmappedPath =>
+            {
+                foreach (var potentialPathMapping in invocationInfo.PathMappings)
+                {
+                    // If it's just a file name being mapped, just a direct map
+                    if (unmappedPath.Equals(potentialPathMapping.From, StringComparison.OrdinalIgnoreCase))
+                    {
+                        return potentialPathMapping.To;
+                    }
+
+                    // Map arbitrary contents under subdirectories
+                    var fromWithDirectorySuffix = AddDirectorySuffixIfMissing(potentialPathMapping.From);
```

Comment:
It's file name to file name or directory to directory. Did that answer your immediate question?

Reply:
eg. say from = filename (C:\foo.cs)
when it hits this line, since C:\foo.cs doesn't end in \, it will append resulting in C:\foo.cs\ , does that matter?

Reply:
The file name case is handled right before "if this is just a file name being mapped...", so we won't even get here.